src/IO/WriteBufferFromS3.cpp

```diff
@@ -182,6 +182,17 @@ void WriteBufferFromS3::finalizeImpl()
     if (!is_prefinalized)
         preFinalize();
 
+    if (std::uncaught_exceptions())
```

Comment:
I don't understand this `if`. So it means that we already have an exception which is not caught yet. And looks like we can get into this `if` only in destructor of some object. In this case we will get `std::terminate` because exceptions are not allowed in destructor (you will get two simultaneous unprocessed exceptions).

Reply:
That is tricky part. Which requires a comment in the code will add one.

So my intention first:
I want to hight light in the tests the cases when finalize are called from d-tors.
In general it is not possible. But It is possible to hight light cases when `finalize` is called when the stack is unwinding. 

So the following explanation is about that specific case.

When stack is unwinding d-tors are called. Well, mainly. Also there could be the code in catches sections. For the catches, I do not see the reason to call finalize for buffers in it, just destroy the instance.

The exceptions could happen in `WriteBufferFromS3` or in the other part of code.
If exception is thrown in the other part of code. It is wrong to call `finalize` for buffer, but at least it is not fatal action. An useless object would be written and abandoned. This part of code hight light such cases.
If exceptions is thrown in the `WriteBufferFromS3` code, than 2 more cases: it could be single part upload or multi. If single then `WriteBufferFromS3`  tries to write it one more time. The consequences are similar to the previous case. If it was multi part upload and finalize is called from the d-tors it is much worse. The call will try to finalize the object and will fail in one of the check here
https://github.com/ClickHouse/ClickHouse/blob/f2172ab8dddbde33d6155e5d991c48818c9d21dc/src/IO/WriteBufferFromS3.cpp#L522-L533
As a result: termination in debug and double exception in release. But usually the call `finalize` is done inside muted exception scope. So nothing happens in release.

So. Lets me conclude:
- nothing for the release build
- makes single part upload behave as a multipart already do in debug buil: termination
So with this code I hope to catch more interesting cases.

And finally, I could remove it if looks too suspicious. It is always possible to put such traps into the designated PR temporary.